### ISSUE TITLE
Inputs with output_differentiability=False don't bump version counter

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#14571]
Example (from autogenerated code):
```
 - name: no_grad_embedding_renorm_(Tensor self, Tensor indices, double max_norm, double norm_type)
   output_differentiability: [False, False, False, False]
```
```
Tensor & VariableType::no_grad_embedding_renorm_(Tensor & self, const Tensor & indices, double ma
x_norm, double norm_type) const {
  profiler::RecordFunction profiler("no_grad_embedding_renorm_", Function::peek_at_next_sequence_
nr());
  auto& self_ = unpack(self, "self", 0);
  auto& indices_ = unpack(indices, "indices", 1);
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    if (tracer_state->force_outplace) {
      op_name = jit::Symbol::fromQualString("aten::no_grad_embedding_renorm");
    } else {
      op_name = jit::Symbol::fromQualString("aten::no_grad_embedding_renorm_");
    }
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "indices", indices);
    jit::tracer::addInputs(node, "max_norm", max_norm);
    jit::tracer::addInputs(node, "norm_type", norm_type);
    tracer_state->graph->appendNode(node);
    jit::tracer::ensureUniqueIfOutOfPlaced("no_grad_embedding_renorm_", self);
    jit::tracer::setTracingState(nullptr);
  }
  baseType->no_grad_embedding_renorm_(self_, indices_, max_norm, norm_type);
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  return self;
}
```

We don't bump the version counter of `self` even though it is modified in place. This sounds dangerous; we could have used `self` in another computation where it is being saved for some Function's backward pass.

cc @colesbury @yf225

**Comment by eellison** [comment]
closed with https://github.com/pytorch/pytorch/pull/14639

**Comment by zou3519** [comment]
Let's still try to figure out what the behavior of output_differentiability should be. As I mentioned in #14633,

> If we only use output_differentiability=False on tensors that are never exposed to the user, then we don't have a problem unless a pytorch dev modifies a tensor in place. If we say that output_differentiability should be user-facing, then I agree we should fix the codegen.

**Comment by ezyang** [comment]
putting @yf225 as he is touching version counters

**Comment by gchanan** [comment]
@nairbv said he'd look at this.

**Comment by nairbv** [comment]
`no_grad_embedding_renorm_` was removed from derivatives.yaml in  https://github.com/pytorch/pytorch/pull/14639/. Do we have some other test-case we can use to observe/test the behavior in the current codebase?

It seems something like the below would work, but no VariableType_*.cpp files were affected when I tried this change:

```
diff --git a/tools/autograd/gen_variable_type.py b/tools/autograd/gen_variable_type.py
index 41f2c1453..f5b3ff0fb 100644
--- a/tools/autograd/gen_variable_type.py
+++ b/tools/autograd/gen_variable_type.py
@@ -519,6 +519,7 @@ def emit_body(declaration):
             if differentiable:
                 differentiable_outputs.append(output)
     elif uses_single_grad(func):
+        candidate_differentiable_outputs = candidate_differentiable_outputs[:1]
         differentiable_outputs = candidate_differentiable_outputs[:1]
     else:
         differentiable_outputs = candidate_differentiable_outputs
@@ -848,7 +849,7 @@ def emit_body(declaration):
     def emit_increment_version():
         if not modifies_arguments:
             return []
-        return ['increment_version({});'.format(arg['name']) for arg in differentiable_outputs]
+        return ['increment_version({});'.format(arg['name']) for arg in candidate_differentiable_outputs]

     def check_record_function_input_type(simple_type):
         return simple_type in ['Tensor', 'Scalar']
```

**Comment by nairbv** [comment]
I think I need some feedback on expected behavior. Tests fail in the PR but I'm not sure if those tests are valid ... there are tests that I think are modifying grad inputs that they shouldn't be modifying.

**Comment by zou3519** [comment]
Closed via #20612
